### PR TITLE
[DataGridPro] Add column pinning

### LINKED ISSUE #193: [DataGridPro] Implement Column pinning
### Benchmark

- https://ant.design/components/table/#components-table-demo-fixed-columns
- https://devexpress.github.io/devextreme-reactive/react/grid/docs/guides/fixed-columns/
- https://ag-grid.com/javascript-grid-pinning/
- https://www.telerik.com/kendo-react-ui/components/grid/columns/locked/
- https://demos.telerik.com/kendo-ui/grid/frozen-columns
- http://schrodinger.github.io/fixed-data-table-2/fixed-right-columns.html
- https://ej2.syncfusion.com/react/demos/#/material/grid/frozen-rows
- http://tabulator.info/docs/4.5/layout#frozen
- https://www.jqwidgets.com/react/react-grid/#https://www.jqwidgets.com/react/react-grid/react-grid-pinnedfrozencolumns.htm
- https://material-table.com/#/docs/features/fixed-columns, the issue: https://github.com/mbrn/material-table/issues/43

### PR DESCRIPTION
**Pull request by m4theushw**
Closes #193 

Preview: https://deploy-preview-2946--material-ui-x.netlify.app/components/data-grid/columns/#column-pinning

Preview: https://deploy-preview-2946--material-ui-x.netlify.app/components/data-grid/demo/

### How it was implemented

The virtualization is composed of 3 components: the viewport, the content and the render zone, where the rows are actually rendered. The render zone is also the part which is moved horizontally and vertically when the user scroll a certain amount. To render a determined set of rows and columns, a `RenderContext` object is given to `getRows` with the indexes to render. The idea behind the column pinning is to make the render zone to never render a pinned column, while the pinned columns are rendered outside of it with `position: sticky`. To achieve that, I changed the main render context to begin rendering after the last left pinned column and to end rendering before the first right pinned column. Between the render zone, two render contexts are created to render only the pinned columns. One problem of this approach is that for each visible row, there're can be 3 `.MuiDataGrid-row` elements, so the hover styles doesn't work properly. To fix that, I'm manually adding the `.Mui-hovered` class to the sibling row elements to make them appear as part of one single row.

### Props added

- `pinnedColumns`: controls which columns are pinned. It's an object of type: `{ left?: string[]; right?: string[] }`.
- `onPinnedColumnsChange`: fired whenever a column is pinned or unpinned. It receives an object containing the new pinned columns.
- `disableColumnPinning`: allows to opt-out from the column pinning feature. It removes the buttons from the column menu 
- and throws an error if any API method is called while `true`
- `initialState.pinnedColumns`: allows to initialize the grid with some pinned columns. This prop should be used to avoid having to add another `onPinnedColumnsChange`

### Future work

- Take screenshot of the print preview with pinned columns

### CODE REVIEW

File: packages/grid/_modules_/grid/components/containers/GridMainContainer.tsx
Comment (m4theushw) [inline packages/grid/_modules_/grid/components/containers/GridMainContainer.tsx:29]: Fixes the missing right border in https://deploy-preview-2946--material-ui-x.netlify.app/components/data-grid/columns/#pinning-the-checkbox-selection-column

image

**Comment by flaviendelangle** [comment]
The column resizing works fine for the pinned column on the left but on the right the behavior is broken.

(https://deploy-preview-2946--material-ui-x.netlify.app/components/data-grid/columns/#column-pinning 1st demo)

**Comment by m4theushw** [comment]
@flaviendelangle Why do you think it's broken? I see this as the expected behavior. The last column can also be resized in https://mui.com/components/data-grid/demo/. The difference here is that it's pinned so it grows to the left.

**Comment by flaviendelangle** [comment]
I feel like it's broken because I move my mouse from a few pixel and the column grows by a lot more.

**Comment by flaviendelangle** [comment]
https://user-images.githubusercontent.com/3309670/142166768-39072c5f-c237-41be-9ba3-37c9ddbfe858.mp4

I think the keyboard navigation is somehow broken when we try to go back from the right pinned column to the unpinned by pressing `ArrowRight`

File: packages/grid/x-data-grid-pro/src/DataGridProColumnHeaders.tsx
Comment (flaviendelangle) [inline packages/grid/x-data-grid-pro/src/DataGridProColumnHeaders.tsx:14]: We should add a `_modules_/grid/hooks/features/columnPinning/index.ts` which exports `columnPinningSelector.ts` and `gridColumnPinningState.ts`

File: packages/grid/x-data-grid-pro/src/DataGridProVirtualScroller.tsx
Comment (flaviendelangle) [reply packages/grid/x-data-grid-pro/src/DataGridProVirtualScroller.tsx:49]:
```suggestion
const getOverlayAlpha = (elevation: number) => {
```

**Review by flaviendelangle — COMMENTED** [review]
I will do a more behavior oriented review later

**Comment by alexfauquette** [comment]
About blocking the pinned column, only removing the option in the menu, let you unpin the column by dropping another column in the pinned one

https://user-images.githubusercontent.com/45398769/142387691-08db459b-c2f6-418b-9aad-44fa6d97a13f.mp4

File: packages/grid/_modules_/grid/hooks/features/columnPinning/useGridColumnPinning.tsx
Comment (m4theushw) [reply packages/grid/_modules_/grid/hooks/features/columnPinning/useGridColumnPinning.tsx:56]: I changed to `rowElements`.

File: packages/grid/_modules_/grid/hooks/utils/useGridState.ts
Comment (m4theushw) [reply packages/grid/_modules_/grid/hooks/utils/useGridState.ts:30]: Opsss. At some time, it would be nice to see what is in the state, similar to redux-logger or Redux DevTools.

File: packages/grid/x-data-grid-pro/src/DataGridProColumnHeaders.tsx
Comment (m4theushw) [reply packages/grid/x-data-grid-pro/src/DataGridProColumnHeaders.tsx:14]: Done.

File: packages/grid/_modules_/grid/models/api/gridColumnPinningApi.ts
Comment (m4theushw) [reply packages/grid/_modules_/grid/models/api/gridColumnPinningApi.ts]: I added an enum.

File: packages/grid/x-data-grid-pro/src/DataGridProVirtualScroller.tsx
Comment (m4theushw) [reply packages/grid/x-data-grid-pro/src/DataGridProVirtualScroller.tsx]: Thanks.

File: packages/grid/_modules_/grid/hooks/features/columnPinning/useGridColumnPinning.tsx
Comment (m4theushw) [reply packages/grid/_modules_/grid/hooks/features/columnPinning/useGridColumnPinning.tsx:50]: Description added.

File: packages/grid/_modules_/grid/hooks/features/columnPinning/useGridColumnPinning.tsx
Comment (m4theushw) [reply packages/grid/_modules_/grid/hooks/features/columnPinning/useGridColumnPinning.tsx]: It doesn't make much difference, it's only the default value. I changed to `undefined` to make clear that it can have this value.

File: scripts/exportsSnapshot.json
Comment (m4theushw) [reply scripts/exportsSnapshot.json]: Users can't use these components today since there's no slot. Anyway, I reverted the change.

File: packages/grid/_modules_/grid/hooks/utils/useGridState.ts
Comment (flaviendelangle) [reply packages/grid/_modules_/grid/hooks/utils/useGridState.ts]: Yes, it should handle multi-DataGrid but it would be nice I agree

File: scripts/exportsSnapshot.json
Comment (flaviendelangle) [reply scripts/exportsSnapshot.json]: Theoretically they can use them in one of there slot, even if I agree that it's unlikely.

**Comment by m4theushw** [comment]
> About blocking the pinned column, only removing the option in the menu, let you unpin the column by dropping another column in the pinned one

@alexfauquette Fixed.

> The column grows very fast, then disappear, then reappears when I mouse up.

@flaviendelangle The problem comes from the fact that the right pinned columns grow to the left while we also use the left position to calculate the new width, so one backfeeds the other and it appears as growing exponentially. I propose to fix this once this PR is merged. Doing it now involves allowing to the change the position of the column separator, which is always on the right size now.

https://github.com/mui-org/material-ui-x/blob/d826fc758ee01cb2a599042f7d19cfc173a123a9/packages/grid/_modules_/grid/hooks/features/columnResize/useGridColumnResize.tsx#L143

**Comment by flaviendelangle** [comment]
You decided to put the `pinnedColumns` in a standalone prop instead of a `pinned` property of `GridColDef`.
I understand the advantage for controlling it. For the grouping I followed AG-Grid which is using a property on the column definition. But we may want to unify the two features.

Anyway, a downside of the standalone prop is that it is possible to have columns in `pinnedColumns` which do not exist.
And the behavior should remain correct.

I noticed a small bug, if you have an invalid column, you can re-order correctly because the column reordering pre-processing uses the amount of pinned columns without sanitizing them.

https://codesandbox.io/s/datagridpro-v5-quick-start-forked-t50nq?file=/package.json
Try to set the "Age" column left of the "Name" column.

File: packages/grid/_modules_/grid/hooks/features/columnPinning/useGridColumnPinning.tsx
Comment (flaviendelangle) [inline packages/grid/_modules_/grid/hooks/features/columnPinning/useGridColumnPinning.tsx:56]: This one crashes if `left` is defined but not `right` or the opposite.
You could replace both `if` statement with

```suggestion
      if ((pinnedColumns.left?.length ?? 0 + pinnedColumns.right?.length ?? 0) === 0) {
```

**Comment by flaviendelangle** [comment]
The feature seems broken when we have horizontal scroll

Example with pinned right and not enough space to display all the non-pinned columns:
https://codesandbox.io/s/datagridpro-v5-quick-start-forked-o7txr

Example with pinned left and not enough space to display all the non-pinned columns:
https://codesandbox.io/s/datagridpro-v5-quick-start-forked-xztni

Example with pinned left and not enough space to display all the pinned columns:
https://codesandbox.io/s/datagridpro-v5-quick-start-forked-yewte
For that one I don't know what the expected behavior is. RIght now we have a desync between the rows and the header.

File: packages/grid/_modules_/grid/hooks/features/columnPinning/useGridColumnPinning.tsx
Comment (m4theushw) [reply packages/grid/_modules_/grid/hooks/features/columnPinning/useGridColumnPinning.tsx]: Fixed.

**Comment by m4theushw** [comment]
About https://github.com/mui-org/material-ui-x/pull/2946#issuecomment-979810240

@flaviendelangle material-table uses a standalone prop: https://material-table.com/#/docs/features/fixed-columns I used the same approach because it reduces the noise in the column definition for those users that use the MIT version. Also, if the pinned flag goes to the column instead, users will have to listen for `onPinnedColumnsChange` but update a prop on each column. 

For the row grouping feature, if there's no button inside the grid for the user to change which column to group - it works in one way only -, then is fine to have it in the column definition. In the column pinning, however, there're buttons to pin columns, it works in two way. 

> Anyway, a downside of the standalone prop is that it is possible to have columns in pinnedColumns which do not exist.
And the behavior should remain correct.

Yeah, we have to always sanitize the value before using. The same for every other controlled prop.

About https://github.com/mui-org/material-ui-x/pull/2946#issuecomment-979818581

All of these examples touch on a limitation of the pinned columns: the grid should be large enough to accommodate them. The same issue also occurs with a div sticky to the right: https://codesandbox.io/s/autumn-wave-dnkdc?file=/src/App.js When there's no space available, it ignores `right: 0`. The column headers are working correctly because they don't use `position: sticky`. I've added a warning to the docs.

**Comment by flaviendelangle** [comment]
> For the row grouping feature, if there's no button inside the grid for the user to change which column to group

There will be the drag&drop pannel.
So I think I'll do like you and add a `groupedColumns` / `onGroupedColumnsChange` instead of following AG-Grid approach. At leat I'll test to see what it would look like.

**Comment by flaviendelangle** [comment]
If you switch `disableColumnPinning` from `true` to `false` it does not stop pinning the column: https://codesandbox.io/s/datagridpro-v5-quick-start-forked-fyqbe


If you pin a column both left and right, it pin it left and pin another random one right (the last one I suppose): https://codesandbox.io/s/datagridpro-v5-quick-start-forked-e6toc
I think it would be nice to at least warn when a column is registered in both.  This clearly is an edge case and can be resolved later.

If you change the position of the pinned column it pins another-one and crashes: https://codesandbox.io/s/datagridpro-v5-quick-start-forked-f9m8z
Not sure what the correct behavior is here. Should we throw when trying to set the index of a pinned column ?

Otherwise, I tried playing with the pinned columns (hide them / change their position / ...) and it behave fine.

**Comment by m4theushw** [comment]
> If you switch disableColumnPinning from true to false it does not stop pinning the column: https://codesandbox.io/s/datagridpro-v5-quick-start-forked-fyqbe

The column pinning, different from the tree data, is opt-out. This prop is meant to allow users to completely disable the column pinning and have the grid as it was before this feature. I didn't invest time into making it to react to changes, it was created to be set initially and never changed.

> If you pin a column both left and right, it pin it left and pin another random one right (the last one I suppose): https://codesandbox.io/s/datagridpro-v5-quick-start-forked-e6toc
I think it would be nice to at least warn when a column is registered in both. This clearly is an edge case and can be resolved later.

Fixed.

> If you change the position of the pinned column it pins another-one and crashes: https://codesandbox.io/s/datagridpro-v5-quick-start-forked-f9m8z

It was crashing when hovering a row, but this was because your CodeSandbox was using old code. Here's the updated one: https://codesandbox.io/s/datagridpro-v5-quick-start-forked-f9m8z?file=/src/App.tsx

I didn't fix the behavior of `setColumnIndex` because it's on the same columns hook and I don't want to put code related to column pinning there. Once this method is moved to a hook dedicated to column reorder I'll check there.

**Comment by flaviendelangle** [comment]
It's good for me
I'll approve once the CI is green :+1:

File: packages/grid/_modules_/grid/hooks/features/columnPinning/useGridColumnPinning.tsx
Comment (flaviendelangle) [inline packages/grid/_modules_/grid/hooks/features/columnPinning/useGridColumnPinning.tsx:137]: Could we type `initialValue` here
I'll do a PR to correctly type the pre-processors one day :+1:

File: packages/grid/_modules_/grid/hooks/features/columnPinning/useGridColumnPinning.tsx
Comment (flaviendelangle) [inline packages/grid/_modules_/grid/hooks/features/columnPinning/useGridColumnPinning.tsx:138]:
```suggestion
    (initialValue: JSX.Element[]) => {
```

Could we type `initialValue` here ?
I'll do a PR to correctly type the pre-processors one day :+1:

Comment (m4theushw) [reply packages/grid/_modules_/grid/hooks/features/columnPinning/useGridColumnPinning.tsx]: Done.

**Comment by m4theushw** [comment]
Because of #3264 changed the argument passed to the `hydrateColumns` pre-processors, I added new `gridVisibleColumnFieldsSelector` selector to get only the field names of the visible columns.

**Review by flaviendelangle — APPROVED** [review]

**Comment by DanailH** [comment]
Demo for the release:
ezgif-3-13a2abdfb0b3

**Pull request merged**
